smarc_bt/src/mission_plan.py

```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+                dubins_waypoints.append(el[ind]) # A row of path[i] represents a single point
+
+        # Sort the waypoints
+        full_path = []
+        ordered_idxs = []
+        for el in path:
+            for e in el:
+                full_path.append([e[0], e[1], e[2]])
+        for j, el in enumerate(dubins_waypoints):
+            try:
+                ordered_idxs.append(full_path.index([el[0], el[1], el[2]])) # Get the dubins wp index in the full path
+            except ValueError:
+                rospy.logwarn("Point " + str(j) + " not found in the path!")
+        ordered_idxs = np.sort(ordered_idxs)
+        dubins_waypoints_ordered = [full_path[i] for i in ordered_idxs]
+
+        # Include original waypoints
+        # The first one is lolo's position, we don't need it
+        wp_i = num_points
+        for wp in waypoints_complete[1:-1]:
+            dubins_waypoints_ordered.insert(wp_i, [wp.x, wp.y, wp.psi])
+            wp_i += num_points + 1
+        dubins_waypoints_ordered.append([waypoints_complete[-1].x, waypoints_complete[-1].y, waypoints_complete[-1].psi])
+        dubins_waypoints_final = dubins_waypoints_ordered[1:]
+
+        dubins_mission = MissionControl()
+        dubins_mission = mission
+        del dubins_mission.waypoints[:]
+
+        # utm_to_latlon service called here to avoid calling it in the loop
+        utm_to_ll_serv = self._get_utm_to_latlon_service()
+
+        k = 0
+        for wp in dubins_waypoints_final:
+            dwp = GotoWaypoint()
+            dwp.pose.pose.position.x = wp[0]
+            dwp.pose.pose.position.y = wp[1]
+            dwp.pose.pose.position.z = travel_altitude
+            dwp.lat, dwp.lon = self.utm_to_latlon(wp[0], wp[1], utm_to_ll_serv)
+
+            quaternion = quaternion_from_euler(0.0, 0.0, math.radians(wp[2])) # RPY [rad]
+            # Normalize quaternion
+            if len(quaternion) != 0:
+                quaternion = quaternion / np.sqrt(np.sum(quaternion**2))
+            dwp.pose.pose.orientation.x = quaternion[0]
+            dwp.pose.pose.orientation.y = quaternion[1]
+            dwp.pose.pose.orientation.z = quaternion[2]
+            dwp.pose.pose.orientation.w = quaternion[3]
+
+            dwp.goal_tolerance = goal_tolerance
+            dwp.z_control_mode = z_control_mode
+            dwp.travel_altitude = travel_altitude 
+            dwp.travel_depth = travel_depth
+            dwp.speed_control_mode = speed_control_mode
+
+            dwp.travel_speed = travel_speed
+            # Speed up for the first dubins waypoint in each segment and then slow down for the following ones
+            # if k % (num_points+1) == 0: 
+            #     dwp.travel_rpm = travel_rpm
+            #     dwp.travel_speed = travel_speed               
+            # else:    
+            #     dwp.travel_rpm = travel_rpm / 1.5
+            #     dwp.travel_speed = travel_speed / 1.5
+            
+            dwp.name = "dwp" + str(k)
```

Comment:
Would be nice if you could somehow reference the two WPs this dwp is in between in its name, so that we have some idea where it is going without a graphical interface. IF it isn't too much of a hassle.